lib/imgresizer.js

```diff
@@ -147,64 +154,49 @@ function processImage(params, method, callback) {
 
       imgcache.get_cached_path(filename, function(dest, exists) {
         if (exists && !force) {
-          // see if the converted file is ready to be serverd
-          fs.open(dest, 'r', function(err, fd) {
-            // try to aquire SHARED lock (if can't means the file is still being converted)
-            fs.flock(fd, 'shnb', function(err) {
-              fs.close(fd, function() {
-                if (err) {
-                  log.trace('couldn\'t lock ' + dest + ', seems like the file is still being converted (' + err.message + ')');
-                  return setTimeout(function() {
-                    processImage(params, method, callback);
-                  }, 0); // retry
-                }
-
-                // just get the info of the image (without converting it)
-                imagemagick.info(dest, function(err, info) {
-                  if (err) {
-                    log.error(err.stack);
-                    return callback(err);
-                  }
-
-                  log.verbose('the image "' + path.basename(dest) + '" seems to be already converted and cached, skipping...');
-                  callback(null, dest, info);
-                });
-              });
-            });
+          // We already have a cached transform; just return image info.
+          imagemagick.info(dest, function(err, info) {
+            if (err) {
+              logError(err);
+              callback(err);
+              return;
+            }
+            log.verbose('The image "' + path.basename(dest) + '" is already converted and cached, skipping...');
+            callback(null, dest, info);
           });
         } else {
-          // prepare file for converting, by opening it in write (append) mode
-          fs.open(dest, 'a', function(err, fd) {
-            // try to aquire EXCLUSIVE lock (cause we're writing to file)
-            fs.flock(fd, 'exnb', function(err) {
+
+          // Transform the image; start in working dir and promote to cache dir once transform succeeds.
+          imgcache.get_working_path(filename, function(workingPath, workingExists) {
+            imgcache.ensure_path(workingPath, function(err) {
```

Comment:
I'd originally matched up the callback args returned by get_working_path with what get_cached_path returned, but workingExists isn't directly useful apart from symmetry/diagnostics.

You're correct that I could have collapsed ensure_path into both get_cached_path and get_working_path. I didn't because that would have meant calls to either get_xxx_path routine would need to handle errors, so I left the read-only code paths the same and only ensuring (and handling possible errors) on the code paths that end up writing files.